src/bitmap.rs

```diff
@@ -0,0 +1,85 @@
+//! This module holds abstractions that enable tracking the areas dirtied by writes of a specified
+//! length to a given offset. In particular, this is used to track write accesses within a
+//! `GuestMemoryRegion` object, and the resulting bitmaps can then be aggregated to build the
+//! global view for the entire `GuestMemory`.
+
+use std::fmt::{self, Debug};
+
+/// Common bitmap operations.
+// This is just an initial proposal; we'll add other methods if they need to be part of the main
+// interface as well.
+pub trait Bitmap: Default {
+    // For now this is mostly useful to preserve compatibility with the old `GuestMemoryMmap` (and
+    // potentially other backends) constructors, which are not bitmap-aware. We can remove this
+    // from the interface (and apply changes elsewhere) if it's too restrictive and/or configuring
+    // the bitmap post-creation (via `GuestMemoryRegion::bitmap_mut`) gets unwieldy.
+    /// Create a new bitmap of the given size.
+    fn new(size: usize) -> Self;
+
+    /// Mark the memory range specified by the given `offset` and `len` as dirtied.
+    fn mark_dirty(&self, offset: usize, len: usize);
+}
+
+/// A no-op `Bitmap` implementation that can be provided to backends that do not actually
+/// require the tracking functionality.
+impl Bitmap for () {
+    fn new(_size: usize) -> Self {
+        ()
+    }
+
+    fn mark_dirty(&self, _offset: usize, _len: usize) {}
+}
+
+/// Represents a slice into a `Bitmap` object, starting at `base_offset`.
+pub struct BitmapSlice<'a, B> {
+    inner: &'a B,
+    base_offset: usize,
+}
+
+impl<'a, B: Bitmap> BitmapSlice<'a, B> {
+    /// Create a new `BitmapSlice`, starting at the specified `offset`.
+    pub fn new(bitmap: &'a B, offset: usize) -> Self {
+        BitmapSlice {
+            inner: bitmap,
+            base_offset: offset,
+        }
+    }
+
+    /// Mark the memory range specified by the given `offset` (relative to the base offset of
+    /// the slice) and `len` as dirtied.
+    pub fn mark_dirty(&self, offset: usize, len: usize) {
+        // The `Bitmap` operations are supposed to accompany guest memory accesses defined by the
+        // same parameters (i.e. offset & length), so we use simple wrapping arithmetic instead of
+        // performing additional checks. If an overflow would occur, we simply end up marking some
+        // other region as dirty (which is just a false positive) instead of a region that could
+        // not have been accessed to begin with.
+        self.inner
+            .mark_dirty(self.base_offset.wrapping_add(offset), len)
```

Comment:
a normal Add instead of wrapping_add() here should be OK, right?
After all we don't support address wrap around for a region.